haystack/utils/reflection.py

```diff
@@ -33,3 +37,47 @@ def wrapper_invocation_counter(*args, **kwargs):
 
     wrapper_invocation_counter.counter = 0
     return wrapper_invocation_counter
+
+
+def retry_with_exponential_backoff(
+    initial_delay: float = 1,
+    exponential_base: float = 2,
+    jitter: bool = True,
+    max_retries: int = 10,
+    errors: tuple = (OpenAIRateLimitError,),
+):
+    """Retry a function with exponential backoff."""
+
+    def decorator(function):
+        def wrapper(*args, **kwargs):
+            # Initialize variables
+            num_retries = 0
+            delay = initial_delay
+
+            # Loop until a successful response or max_retries is hit or an exception is raised
+            while True:
+                try:
+                    return function(*args, **kwargs)
+
+                # Retry on specified errors
+                except errors as e:
+                    # Increment retries
+                    num_retries += 1
+
+                    # Check if max retries has been reached
+                    if num_retries > max_retries:
+                        raise Exception(f"Maximum number of retries ({max_retries}) exceeded.")
+
+                    # Increment the delay
+                    delay *= exponential_base * (1 + jitter * random())
```

Comment:
Hey @vblagoje, I think exponential backoff with jitter is usually implemented like 
```python
((2^n)+random_number_milliseconds)
```
where `n` is the number of retries. 

Otherwise using the above algorithm we have a situation where the jitter is changing the exponential base which can result in longer than anticipated wait times. What do you think?

Reply:
TBH I copied this code from the OpenAI example on how to do back off. I'll add the unit test you suggested and think this through again.